specs/plasma.md

```diff
@@ -0,0 +1,114 @@
+commitment (keccak256 hash) is submitted as the bacher inbox transaction call data.
+The batcher will not submit a commitment onchain unless input data was successfully stored on the service.
+
+## Data Availability Challenge Contract
+
+Data availability is guaranteed via a permissionless challenge contract on the L1 chain.
+Users have a set number of L1 blocks (AKA `challenge_window`) during which they are able to call
+the `challenge` method of the contract with the following inputs:
+
+- A Commitment type (i.e. keccak256)
+- Commitment bytes
+- The L1 block number in which it was included.
+
+Users with access to the input data then have another window of L1 blocks (AKA `resolve_window`)
+during which they can submit it as calldata to the chain by calling the `resolve` method of the contract.
+If the data is not included onchain by the time the resolve window is elapsed, derivation of the L2 canonical chain
+will omit the input data, hence any transaction included in the input data will be dropped
+causing a reorg of any L2 chain that previously had included that transaction.
+
+In order to challenge a commitment, users must deposit a bond amount where `bond >= resolve_tx_gas_cost`.
+If the gas cost of resolving the challenge was lower than the bond, the difference is reimbursed to the challenger
+and the rest of the bond is burnt. If the challenge is not resolved in time and expired,
+the bond is returned and can be withdrawn by the challenger or used to challenge another commitment.
+
+The state of all challenges can be read from the contract state or by syncing contract events.
+`challenge_window` and `resolve_window` are constant values that currently cannot be changed
+unless the contract is upgraded. A dynamic window mechanism may be explored in the future.
+
+## Derivation
+
+Input data is retrieved during derivation of L2 blocks. The changes to the derivation pipeline
+when using the alt DA source are limited to swapping out the L1 based `DataAvailabilitySource` step
+in the pipeline with a module that enables pulling the data from the offchain DA source.
+
+Similarly to L1 based DA, for each L1 block we open a calldata source to retrieve the input commitments
+from the transactions and use each commitment with its l1 origin block number to resolve
+the input data from the storage service.
+
+In addition, we filter events from the DA Challenge contract included in the block
+and sync a local state of challenged input commitments. As the derivation pipeline steps through
+`challenge_window + resolve_window` amount of L1 blocks, any challenge marked as active
+becomes expired causing a reset of the derivation pipeline.
+
+Derivation can either be driven by new L1 blocks or restarted (reset) from the L1 origin of the last
+L2 safe head known by the execution engine.
+
+When stepping through new L1 origin blocks, input data is loaded from the DA storage service.
```

Comment:
This additional data input into the derivation pipeline will need to be integrated with fault proofs as well by adding an additional [hint type for the preimage oracle](https://github.com/ethereum-optimism/specs/blob/main/specs/fault-proof.md#pre-image-hinting-routes). The hint provides context to a host program when running the fault proof derivation off-chain, so it knows what the hash about to be requested represents.

Reply:
Could we add a brief section on the fault proof or modify the hints list to capture this?